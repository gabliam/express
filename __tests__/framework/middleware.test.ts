import {
  Controller,
  Get,
  Post,
  Put,
  Patch,
  Head,
  Delete,
  All,
  ExpressConfig,
  Middleware,
  MiddlewareInject
} from '../../src/index';
import * as e from 'express';
import { ExpressPluginTest } from '../express-plugin-test';
import * as supertest from 'supertest';
import { Config } from '@gabliam/core';
import * as sinon from 'sinon';

let appTest: ExpressPluginTest;

beforeEach(async () => {
  appTest = new ExpressPluginTest();
});

afterEach(async () => {
  await appTest.destroy();
});

describe('Middleware:', () => {
  let result: string;
  const middleware: any = {
    a: function(req: e.Request, res: e.Response, nextFunc: e.NextFunction) {
      result += 'a';
      nextFunc();
    },
    b: function(req: e.Request, res: e.Response, nextFunc: e.NextFunction) {
      result += 'b';
      nextFunc();
    },
    c: function(req: e.Request, res: e.Response, nextFunc: e.NextFunction) {
      result += 'c';
      nextFunc();
    }
  };
  const spyA = sinon.spy(middleware, 'a');
  const spyB = sinon.spy(middleware, 'b');
  const spyC = sinon.spy(middleware, 'c');

  beforeEach(() => {
    result = '';
    spyA.reset();
    spyB.reset();
    spyC.reset();
  });

  test('should call method-level middleware correctly (GET)', async () => {
    @Controller('/')
    class TestController {
      @Get('/', spyA, spyB, spyC)
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
    }
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);

    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (POST)', async () => {
    @Controller('/')
    class TestController {
      @Post('/', spyA, spyB, spyC)
      public postTest(req: e.Request, res: e.Response) {
        res.send('POST');
      }
    }

    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).post('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (PUT)', async () => {
    @Controller('/')
    class TestController {
      @Put('/', spyA, spyB, spyC)
      public postTest(req: e.Request, res: e.Response) {
        res.send('PUT');
      }
    }
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).put('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (PATCH)', async () => {
    @Controller('/')
    class TestController {
      @Patch('/', spyA, spyB, spyC)
      public postTest(req: e.Request, res: e.Response) {
        res.send('PATCH');
      }
    }

    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).patch('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (HEAD)', async () => {
    @Controller('/')
    class TestController {
      @Head('/', spyA, spyB, spyC)
      public postTest(req: e.Request, res: e.Response) {
        res.send('HEAD');
      }
    }
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).head('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (DELETE)', async () => {
    @Controller('/')
    class TestController {
      @Delete('/', spyA, spyB, spyC)
      public postTest(req: e.Request, res: e.Response) {
        res.send('DELETE');
      }
    }
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).delete('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (ALL)', async () => {
    @Controller('/')
    class TestController {
      @All('/', spyA, spyB, spyC)
      public postTest(req: e.Request, res: e.Response) {
        res.send('ALL');
      }
    }
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call controller-level middleware correctly', async () => {
    @Controller({
      path: '/',
      middlewares: [spyA, spyB, spyC]
    })
    class TestController {
      @Get('/')
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
    }

    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call server-level middleware correctly', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
    }

    @Config()
    class ServerConfig {
      @ExpressConfig()
      serverConfig(app: e.Application) {
        app.use(spyA);
        app.use(spyB);
        app.use(spyC);
      }
    }

    appTest.addClass(TestController);
    appTest.addClass(ServerConfig);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call all middleware in correct order', async () => {
    @Controller({
      path: '/',
      middlewares: [spyB]
    })
    class TestController {
      @Get('/', spyC)
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
    }

    @Config()
    class ServerConfig {
      @ExpressConfig()
      serverConfig(app: e.Application) {
        app.use(spyA);
      }
    }

    appTest.addClass(TestController);
    appTest.addClass(ServerConfig);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should resolve controller-level middleware', async () => {
    const symbolId = Symbol('spyA');
    const strId = 'spyB';

    @Controller({
      path: '/',
      middlewares: [symbolId, strId]
    })
    class TestController {
      @Get('/')
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
    }

    appTest.addClass(TestController);
    appTest.gab.container
      .bind<e.RequestHandler>(symbolId)
      .toConstantValue(spyA);
    appTest.gab.container.bind<e.RequestHandler>(strId).toConstantValue(spyB);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should resolve method-level middleware', async () => {
    const symbolId = Symbol('spyA');
    const strId = 'spyB';

    @Controller('/')
    class TestController {
      @Get('/', symbolId, strId)
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
    }

    appTest.addClass(TestController);
    appTest.gab.container
      .bind<e.RequestHandler>(symbolId)
      .toConstantValue(spyA);
    appTest.gab.container.bind<e.RequestHandler>(strId).toConstantValue(spyB);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should compose controller- and method-level middleware', async () => {
    const symbolId = Symbol('spyA');
    const strId = 'spyB';

    @Controller({
      path: '/',
      middlewares: [symbolId]
    })
    class TestController {
      @Get('/', strId)
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
    }

    appTest.addClass(TestController);
    appTest.gab.container
      .bind<e.RequestHandler>(symbolId)
      .toConstantValue(spyA);
    appTest.gab.container.bind<e.RequestHandler>(strId).toConstantValue(spyB);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });
});

describe('Middleware inject:', () => {
  let result: string;
  let args: string;
  @Config()
  class TestConfig {
    @Middleware('a')
    a() {
      return () => (
        req: e.Request,
        res: e.Response,
        nextFunc: e.NextFunction
      ) => {
        result += 'a';
        nextFunc();
      };
    }

    @Middleware('b')
    b() {
      return () => (
        req: e.Request,
        res: e.Response,
        nextFunc: e.NextFunction
      ) => {
        result += 'b';
        nextFunc();
      };
    }

    @Middleware('c')
    c(arg?: string) {
      return () => (
        req: e.Request,
        res: e.Response,
        nextFunc: e.NextFunction
      ) => {
        result += 'c';
        if (arg) {
          args += arg;
        }
        nextFunc();
      };
    }

    @Middleware('de')
    de() {
      return (arg?: string, arg2?: string) => [
        (req: e.Request, res: e.Response, nextFunc: e.NextFunction) => {
          result += 'd';
          if (arg) {
            args += 'd' + arg;
          }
          if (arg2) {
            args += 'd' + arg2;
          }
          nextFunc();
        },
        (req: e.Request, res: e.Response, nextFunc: e.NextFunction) => {
          result += 'e';
          if (arg) {
            args += 'e' + arg;
          }
          if (arg2) {
            args += 'e' + arg2;
          }
          nextFunc();
        }
      ];
    }
  }

  beforeEach(() => {
    result = '';
    args = '';
  });

  test('should call method-level middleware correctly (GET)', async () => {
    @Controller('/')
    class TestController {
      @MiddlewareInject('de', 'dearg', 'dearg2')
      @MiddlewareInject('c', 'carg')
      @MiddlewareInject('b')
      @MiddlewareInject('a')
      @Get('/')
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
    }

    appTest.addClass(TestConfig);
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (POST)', async () => {
    @Controller('/')
    class TestController {
      @MiddlewareInject('de', 'dearg', 'dearg2')
      @MiddlewareInject('c', 'carg')
      @MiddlewareInject('b')
      @MiddlewareInject('a')
      @Post('/')
      public postTest(req: e.Request, res: e.Response) {
        res.send('POST');
      }
    }

    appTest.addClass(TestConfig);
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).post('/').expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (PUT)', async () => {
    @Controller('/')
    class TestController {
      @MiddlewareInject('de', 'dearg', 'dearg2')
      @MiddlewareInject('c', 'carg')
      @MiddlewareInject('b')
      @MiddlewareInject('a')
      @Put('/')
      public postTest(req: e.Request, res: e.Response) {
        res.send('PUT');
      }
    }

    appTest.addClass(TestConfig);
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).put('/').expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (PATCH)', async () => {
    @Controller('/')
    class TestController {
      @MiddlewareInject('de', 'dearg', 'dearg2')
      @MiddlewareInject('c', 'carg')
      @MiddlewareInject('b')
      @MiddlewareInject('a')
      @Patch('/')
      public postTest(req: e.Request, res: e.Response) {
        res.send('PATCH');
      }
    }

    appTest.addClass(TestConfig);
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).patch('/').expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (HEAD)', async () => {
    @Controller('/')
    class TestController {
      @MiddlewareInject('de', 'dearg', 'dearg2')
      @MiddlewareInject('c', 'carg')
      @MiddlewareInject('b')
      @MiddlewareInject('a')
      @Head('/')
      public postTest(req: e.Request, res: e.Response) {
        res.send('HEAD');
      }
    }
    appTest.addClass(TestConfig);
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).head('/').expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (DELETE)', async () => {
    @Controller('/')
    class TestController {
      @MiddlewareInject('de', 'dearg', 'dearg2')
      @MiddlewareInject('c', 'carg')
      @MiddlewareInject('b')
      @MiddlewareInject('a')
      @Delete('/')
      public postTest(req: e.Request, res: e.Response) {
        res.send('DELETE');
      }
    }

    appTest.addClass(TestConfig);
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).delete('/').expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level middleware correctly (ALL)', async () => {
    @Controller('/')
    class TestController {
      @MiddlewareInject('de', 'dearg', 'dearg2')
      @MiddlewareInject('c', 'carg')
      @MiddlewareInject('b')
      @MiddlewareInject('a')
      @All('/')
      public postTest(req: e.Request, res: e.Response) {
        res.send('ALL');
      }
    }

    appTest.addClass(TestConfig);
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call controller-level middleware correctly', async () => {
    @MiddlewareInject('de', 'dearg', 'dearg2')
    @MiddlewareInject('c', 'carg')
    @MiddlewareInject('b')
    @MiddlewareInject('a')
    @Controller({
      path: '/'
    })
    class TestController {
      @Get('/')
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
    }

    appTest.addClass(TestConfig);
    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call all middleware in correct order', async () => {
    @MiddlewareInject('b')
    @MiddlewareInject('a')
    @Controller({
      path: '/'
    })
    class TestController {
      @MiddlewareInject('de', 'dearg', 'dearg2')
      @MiddlewareInject('c', 'carg')
      @Get('/')
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
    }

    appTest.addClass(TestController);
    appTest.addClass(TestConfig);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });
});
