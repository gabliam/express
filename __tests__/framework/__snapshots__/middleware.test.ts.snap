// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Middleware inject: should call all middleware in correct order 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Middleware inject: should call all middleware in correct order 2`] = `"abcde"`;

exports[`Middleware inject: should call all middleware in correct order 3`] = `"ddeargddearg2edeargedearg2"`;

exports[`Middleware inject: should call controller-level middleware correctly 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Middleware inject: should call controller-level middleware correctly 2`] = `"abcde"`;

exports[`Middleware inject: should call controller-level middleware correctly 3`] = `"ddeargddearg2edeargedearg2"`;

exports[`Middleware inject: should call method-level middleware correctly (ALL) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-a0KHTjzSB3HZMJbsXONjB6HyuhQ\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "ALL",
}
`;

exports[`Middleware inject: should call method-level middleware correctly (ALL) 2`] = `"abcde"`;

exports[`Middleware inject: should call method-level middleware correctly (ALL) 3`] = `"ddeargddearg2edeargedearg2"`;

exports[`Middleware inject: should call method-level middleware correctly (DELETE) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"6-1vVjYJjNRYzp0ik5+OPo3qsOm9A\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "DELETE",
  },
  "status": 200,
  "text": "DELETE",
}
`;

exports[`Middleware inject: should call method-level middleware correctly (DELETE) 2`] = `"abcde"`;

exports[`Middleware inject: should call method-level middleware correctly (DELETE) 3`] = `"ddeargddearg2edeargedearg2"`;

exports[`Middleware inject: should call method-level middleware correctly (GET) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Middleware inject: should call method-level middleware correctly (GET) 2`] = `"abcde"`;

exports[`Middleware inject: should call method-level middleware correctly (GET) 3`] = `"ddeargddearg2edeargedearg2"`;

exports[`Middleware inject: should call method-level middleware correctly (HEAD) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"4-cTilFmGUexm1CI2lor/t4odvSbk\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "HEAD",
  },
  "status": 200,
  "text": undefined,
}
`;

exports[`Middleware inject: should call method-level middleware correctly (HEAD) 2`] = `"abcde"`;

exports[`Middleware inject: should call method-level middleware correctly (HEAD) 3`] = `"ddeargddearg2edeargedearg2"`;

exports[`Middleware inject: should call method-level middleware correctly (PATCH) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"5-n74p1Ndux1ZCGu4G4PNPF/pIj/M\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "PATCH",
  },
  "status": 200,
  "text": "PATCH",
}
`;

exports[`Middleware inject: should call method-level middleware correctly (PATCH) 2`] = `"abcde"`;

exports[`Middleware inject: should call method-level middleware correctly (PATCH) 3`] = `"ddeargddearg2edeargedearg2"`;

exports[`Middleware inject: should call method-level middleware correctly (POST) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"4-Yf+Bwwqjx254r+pisuO9HfpJ6FQ\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "POST",
  },
  "status": 200,
  "text": "POST",
}
`;

exports[`Middleware inject: should call method-level middleware correctly (POST) 2`] = `"abcde"`;

exports[`Middleware inject: should call method-level middleware correctly (POST) 3`] = `"ddeargddearg2edeargedearg2"`;

exports[`Middleware inject: should call method-level middleware correctly (PUT) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-CRsM5C6wvZYWnqALFt2Tj21jrJU\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "PUT",
  },
  "status": 200,
  "text": "PUT",
}
`;

exports[`Middleware inject: should call method-level middleware correctly (PUT) 2`] = `"abcde"`;

exports[`Middleware inject: should call method-level middleware correctly (PUT) 3`] = `"ddeargddearg2edeargedearg2"`;

exports[`Middleware: should call all middleware in correct order 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Middleware: should call all middleware in correct order 2`] = `"abc"`;

exports[`Middleware: should call controller-level middleware correctly 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Middleware: should call controller-level middleware correctly 2`] = `"abc"`;

exports[`Middleware: should call method-level middleware correctly (ALL) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-a0KHTjzSB3HZMJbsXONjB6HyuhQ\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "ALL",
}
`;

exports[`Middleware: should call method-level middleware correctly (ALL) 2`] = `"abc"`;

exports[`Middleware: should call method-level middleware correctly (DELETE) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"6-1vVjYJjNRYzp0ik5+OPo3qsOm9A\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "DELETE",
  },
  "status": 200,
  "text": "DELETE",
}
`;

exports[`Middleware: should call method-level middleware correctly (DELETE) 2`] = `"abc"`;

exports[`Middleware: should call method-level middleware correctly (GET) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Middleware: should call method-level middleware correctly (GET) 2`] = `"abc"`;

exports[`Middleware: should call method-level middleware correctly (HEAD) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"4-cTilFmGUexm1CI2lor/t4odvSbk\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "HEAD",
  },
  "status": 200,
  "text": undefined,
}
`;

exports[`Middleware: should call method-level middleware correctly (HEAD) 2`] = `"abc"`;

exports[`Middleware: should call method-level middleware correctly (PATCH) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"5-n74p1Ndux1ZCGu4G4PNPF/pIj/M\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "PATCH",
  },
  "status": 200,
  "text": "PATCH",
}
`;

exports[`Middleware: should call method-level middleware correctly (PATCH) 2`] = `"abc"`;

exports[`Middleware: should call method-level middleware correctly (POST) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"4-Yf+Bwwqjx254r+pisuO9HfpJ6FQ\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "POST",
  },
  "status": 200,
  "text": "POST",
}
`;

exports[`Middleware: should call method-level middleware correctly (POST) 2`] = `"abc"`;

exports[`Middleware: should call method-level middleware correctly (PUT) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-CRsM5C6wvZYWnqALFt2Tj21jrJU\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "PUT",
  },
  "status": 200,
  "text": "PUT",
}
`;

exports[`Middleware: should call method-level middleware correctly (PUT) 2`] = `"abc"`;

exports[`Middleware: should call server-level middleware correctly 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Middleware: should call server-level middleware correctly 2`] = `"abc"`;

exports[`Middleware: should compose controller- and method-level middleware 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Middleware: should compose controller- and method-level middleware 2`] = `"ab"`;

exports[`Middleware: should resolve controller-level middleware 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Middleware: should resolve controller-level middleware 2`] = `"ab"`;

exports[`Middleware: should resolve method-level middleware 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Middleware: should resolve method-level middleware 2`] = `"ab"`;
